Adafruit_DAP.cpp

```diff
@@ -45,6 +45,26 @@ bool Adafruit_DAP::begin(int swclk, int swdio, int nreset, ErrorHandler perr) {
   // return true;
 }
 
+void Adafruit_DAP::targetConnect() {
+  Serial.println("Connecting...");
```

Comment:
> thank you very much of following the previous feedback, this is such a great PR. However, it seems to piggy-back lots of changes that could make it a bit longer to get merged. I really like the high-level `targetConnect()` but not at all with the `programFlash()` and `resetProtectionFuses()`. We should have a clear cut in API between the application and DAP protocol. DAP layer should be as primitive as possible and should not overstretch the flow control that is dictated by application. I am open to any further discussion, feel free to comment if you disagree.
> 
> PS: whenever you think it is ready again, please click on `re-request review` icon so that I could get the notification, thanks again.

Thank you for the review. 
Yeah, I should have committed with more atomic changes, I guess, but I couldn't afford to spend more time on it. The most important changes I think, is the bug fix of the UserRow fuse handling on the SAMD21 (the buggy code destroyed the voltage ref config on my MCU), that the classes defines the size of the units to be flashed (and not allowing the sketched to do that), the reset with extension, which made it necessary to move connection into the library as well, resetting protection fuses at appropriate place and moving of dap_target_prepare into the library, to be called at the right times. 
I have seen a few discussions in forums, where people have unsuccessfully tried to use Adafruit_DAP to reflash bootloader into MCU for recovery purposes, and my guess is that in some situations this is caused by either having BOOTPROT, Region Locks or Security Bit set. People very often have limited knowledge of the technicalities and needs to have a framework that takes care of things for them.